journalpump/senders/google_cloud_logging.py

```diff
@@ -83,6 +83,18 @@ def send_messages(self, *, messages, cursor):
             self._logs.entries().write(body=body).execute()  # pylint: disable=no-member
             self.mark_sent(messages=messages, cursor=cursor)
             return True
+        except HttpError as err:
+            # Retrying a 400 will probably result in the same error
+            # so in this case we skip this LogEntry but still raise
+            # the exception for keeping visibility of the error and
+            # metrics
+            if isinstance(err, HttpError) and err.status_code >= 400 and err.status_code < 500:
+                self.mark_sent(messages=messages, cursor=cursor)
```

Comment:
I aggree that 4XX errors seem unrecoverable, but there is the possibility that the authentication or authorization could be recovered, or updated after some time. Marking the messages sent would move the sender's cursor, and the sender won't try to send those messages again.

Isn't it enough to mark it disconnected (pausing the sender) and not try to connect until some action is taken? Or we can considerably increase the size of the backoff mechanism's retry intervals, and after some number of attempts, we can just stop the sender.

Reply:
> Isn't it enough to mark it disconnected (pausing the sender) and not try to connect until some action is taken?
In the case of a big message probably not, but, as you mentioned, in case of authentication/authorization error that can certainly change after a configuration change.

> Or we can considerably increase the size of the backoff mechanism's retry intervals, and after some number of attempts, we can just stop the sender.

I thought of having a maximum number of attempts, but the logs are transmitted in batches and the operation is atomic, so it could be hard to know which one(s) of the logs is causing problems.

Reply:
I am not sure I will go forward with this approach of handling 4xx though, I have to think about it a bit more